File: lumina/packages/core/src/detection/yawn.ts
/**
 * Yawn detection using Mouth Aspect Ratio (MAR).
 *
 * Similar to EAR (Eye Aspect Ratio) but for mouth.
 * Yawn = mouth open (high MAR) for extended duration.
 *
 * MAR = vertical_opening / horizontal_width
 * Higher MAR = mouth more open.
 */

import { MOUTH_INDICES, YAWN } from './constants';
import { Point2D } from './blink';

// ============================================================================
// Types
// ============================================================================

export interface YawnResult {
  /** Mouth Aspect Ratio (0-1+) */
  mar: number;
  /** MAR above threshold */
  isMouthOpen: boolean;
  /** Mouth open for required duration */
  isYawning: boolean;
  /** How long mouth has been open (ms) */
  yawnDurationMs: number;
  /** Total yawns detected */
  yawnCount: number;
  /** Recently detected a yawn */
  inCooldown: boolean;
}

// ============================================================================
// MAR Calculation
// ============================================================================

/**
 * Calculate Mouth Aspect Ratio.
 *
 * MAR = vertical_opening / horizontal_width
 *
 * @param landmarks MediaPipe face landmarks
 * @returns MAR value (typically 0.1-0.3 closed, 0.5+ open wide)
 */
export function calculateMAR(landmarks: Point2D[]): number {
  const upperLip = landmarks[MOUTH_INDICES.UPPER_LIP];
  const lowerLip = landmarks[MOUTH_INDICES.LOWER_LIP];
  const leftCorner = landmarks[MOUTH_INDICES.LEFT_CORNER];
  const rightCorner = landmarks[MOUTH_INDICES.RIGHT_CORNER];

  // Vertical mouth opening
  const verticalDist = Math.sqrt(
    Math.pow(upperLip.x - lowerLip.x, 2) +
    Math.pow(upperLip.y - lowerLip.y, 2)
  );

  // Horizontal mouth width
  const horizontalDist = Math.sqrt(
    Math.pow(leftCorner.x - rightCorner.x, 2) +
    Math.pow(leftCorner.y - rightCorner.y, 2)
  );

  // Avoid division by zero
  if (horizontalDist < 0.001) {
    return 0;
  }

  return verticalDist / horizontalDist;
}

// ============================================================================
// Yawn Detector
// ============================================================================

/**
 * Detects yawns based on prolonged mouth opening.
 *
 * A yawn is detected when:
 * - MAR exceeds threshold (mouth is open)
 * - Stays open for YAWN_DURATION_MS or longer
 *
 * Usage:
 * ```typescript
 * const detector = new YawnDetector();
 *
 * // In detection loop
 * const result = detector.update(landmarks);
 * if (result.isYawning) {
 *   console.log('Yawn detected!', result.yawnCount);
 * }
 * ```
 */
export class YawnDetector {
  private mouthOpenStart: number | null = null;
  private _yawnCount: number = 0;
  private lastYawnTime: number = 0;
  private yawnTimestamps: number[] = [];

  get yawnCount(): number {
    return this._yawnCount;
  }

  /**
   * Process frame and detect yawns.
   *
   * @param landmarks MediaPipe face landmarks
   * @param timestampMs Current timestamp in milliseconds
   */
  update(landmarks: Point2D[], timestampMs?: number): YawnResult {
    const now = timestampMs ?? performance.now();
    const mar = calculateMAR(landmarks);
    const isMouthOpen = mar > YAWN.MAR_THRESHOLD;

    // Check cooldown
    const inCooldown = (now - this.lastYawnTime) < YAWN.COOLDOWN_MS;

    let yawnDurationMs = 0;
    let isYawning = false;

    if (isMouthOpen && !inCooldown) {
      // Mouth is open
      if (this.mouthOpenStart === null) {
        this.mouthOpenStart = now;
      }

      yawnDurationMs = now - this.mouthOpenStart;

      // Check if qualifies as yawn
      if (yawnDurationMs >= YAWN.DURATION_MS) {
        isYawning = true;

        // Count yawn once (when first detected)
        // Within 100ms of threshold crossing
        if (yawnDurationMs < YAWN.DURATION_MS + 100) {
          this._yawnCount++;
          this.lastYawnTime = now;
          this.yawnTimestamps.push(now);
        }
      }
    } else {
      // Mouth closed or in cooldown
      this.mouthOpenStart = null;
    }

    return {
      mar,
      isMouthOpen,
      isYawning,
      yawnDurationMs,
      yawnCount: this._yawnCount,
      inCooldown,
    };
  }

  /**
   * Get list of yawn timestamps for drowsiness calculation.
   */
  getYawnTimestamps(): number[] {
    return [...this.yawnTimestamps];
  }

  /**
   * Count yawns within a time window.
   *
   * @param windowMs Time window in milliseconds
   * @param currentTimeMs Current time (uses performance.now if not provided)
   */
  getRecentYawns(windowMs: number, currentTimeMs?: number): number {
    const now = currentTimeMs ?? performance.now();
    const cutoff = now - windowMs;
    return this.yawnTimestamps.filter(t => t > cutoff).length;
  }

  /**
   * Reset detector state.
   */
  reset(): void {
    this.mouthOpenStart = null;
    this._yawnCount = 0;
    this.lastYawnTime = 0;
    this.yawnTimestamps = [];
  }
}

File: experiments/robust_blink_v2/yawn.py
"""
Yawn detection using Mouth Aspect Ratio (MAR).

Similar to EAR (Eye Aspect Ratio) but for mouth.
Yawn = mouth open (high MAR) for extended duration.
"""

from dataclasses import dataclass
from typing import Optional
import math
import time

from constants import (
    MOUTH_UPPER_LIP_INDEX,
    MOUTH_LOWER_LIP_INDEX,
    MOUTH_LEFT_CORNER_INDEX,
    MOUTH_RIGHT_CORNER_INDEX,
    MAR_THRESHOLD,
    YAWN_DURATION_MS,
    YAWN_COOLDOWN_MS,
)


@dataclass
class YawnResult:
    """Result of yawn detection."""
    mar: float                  # Mouth Aspect Ratio (0-1+)
    is_mouth_open: bool         # MAR above threshold
    is_yawning: bool            # Mouth open for required duration
    yawn_duration_ms: float     # How long mouth has been open
    yawn_count: int             # Total yawns detected
    in_cooldown: bool           # Recently detected a yawn


def calculate_mar(landmarks: list) -> float:
    """
    Calculate Mouth Aspect Ratio.

    MAR = vertical_opening / horizontal_width

    Higher MAR = mouth more open.

    Args:
        landmarks: MediaPipe face landmarks

    Returns:
        MAR value (typically 0.1-0.3 closed, 0.5+ open wide)
    """
    upper_lip = landmarks[MOUTH_UPPER_LIP_INDEX]
    lower_lip = landmarks[MOUTH_LOWER_LIP_INDEX]
    left_corner = landmarks[MOUTH_LEFT_CORNER_INDEX]
    right_corner = landmarks[MOUTH_RIGHT_CORNER_INDEX]

    # Vertical mouth opening
    vertical_dist = math.sqrt(
        (upper_lip.x - lower_lip.x) ** 2 +
        (upper_lip.y - lower_lip.y) ** 2
    )

    # Horizontal mouth width
    horizontal_dist = math.sqrt(
        (left_corner.x - right_corner.x) ** 2 +
        (left_corner.y - right_corner.y) ** 2
    )

    # Avoid division by zero
    if horizontal_dist < 0.001:
        return 0.0

    return vertical_dist / horizontal_dist


class YawnDetector:
    """
    Detects yawns based on prolonged mouth opening.

    A yawn is detected when:
    - MAR exceeds threshold (mouth is open)
    - Stays open for YAWN_DURATION_MS or longer
    """

    def __init__(self):
        self.mouth_open_start: Optional[float] = None
        self.yawn_count: int = 0
        self.last_yawn_time: float = 0
        self.yawn_timestamps: list = []  # For drowsiness tracking

    def update(self, landmarks: list, timestamp_ms: Optional[float] = None) -> YawnResult:
        """
        Process frame and detect yawns.

        Args:
            landmarks: MediaPipe face landmarks
            timestamp_ms: Current timestamp in milliseconds

        Returns:
            YawnResult with MAR and yawn status
        """
        if timestamp_ms is None:
            timestamp_ms = time.time() * 1000

        mar = calculate_mar(landmarks)
        is_mouth_open = mar > MAR_THRESHOLD

        # Check cooldown
        in_cooldown = (timestamp_ms - self.last_yawn_time) < YAWN_COOLDOWN_MS

        yawn_duration_ms = 0
        is_yawning = False

        if is_mouth_open and not in_cooldown:
            # Mouth is open
            if self.mouth_open_start is None:
                self.mouth_open_start = timestamp_ms

            yawn_duration_ms = timestamp_ms - self.mouth_open_start

            # Check if qualifies as yawn
            if yawn_duration_ms >= YAWN_DURATION_MS:
                is_yawning = True

                # Count yawn once (when first detected)
                if yawn_duration_ms < YAWN_DURATION_MS + 100:  # Within 100ms of threshold
                    self.yawn_count += 1
                    self.last_yawn_time = timestamp_ms
                    self.yawn_timestamps.append(timestamp_ms)

        else:
            # Mouth closed or in cooldown
            self.mouth_open_start = None

        return YawnResult(
            mar=mar,
            is_mouth_open=is_mouth_open,
            is_yawning=is_yawning,
            yawn_duration_ms=yawn_duration_ms,
            yawn_count=self.yawn_count,
            in_cooldown=in_cooldown,
        )

    def get_yawn_timestamps(self) -> list:
        """Get list of yawn timestamps for drowsiness calculation."""
        return self.yawn_timestamps.copy()

    def get_recent_yawns(self, window_ms: float, current_time_ms: Optional[float] = None) -> int:
        """
        Count yawns within a time window.

        Args:
            window_ms: Time window in milliseconds
            current_time_ms: Current time (uses system time if None)

        Returns:
            Number of yawns in the window
        """
        if current_time_ms is None:
            current_time_ms = time.time() * 1000

        cutoff = current_time_ms - window_ms
        return sum(1 for t in self.yawn_timestamps if t > cutoff)

    def reset(self):
        """Reset detector state."""
        self.mouth_open_start = None
        self.yawn_count = 0
        self.last_yawn_time = 0
        self.yawn_timestamps = []
